packages/react-reconciler/src/ReactFiberBeginWork.js

```diff
@@ -246,7 +247,7 @@ function updatePureComponent(
     // Default to shallow comparison
     let compare = Component.compare;
     compare = compare !== null ? compare : shallowEqual;
-    if (compare(prevProps, nextProps)) {
+    if (workInProgress.ref === current.ref && compare(prevProps, nextProps)) {
```

Comment:
Thanks! Need to fix this for `forwardRef`, too, but we can do that in a separate PR.

Reply:
Looks like we do already:

https://github.com/facebook/react/blob/0af81997095dfa3db67a0c5d287b144797742e99/packages/react-reconciler/src/ReactFiberBeginWork.js#L187-L208

though I'm not sure we need to check legacy context there?

Reply:
Problem is we never hit this code path because of: https://github.com/facebook/react/blob/0af81997095dfa3db67a0c5d287b144797742e99/packages/react-reconciler/src/ReactFiberBeginWork.js#L1209-L1217